forkmanager/fork_manager.go

```diff
@@ -0,0 +1,233 @@
+package forkmanager
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+/*
+Regarding whether it is okay to use the Singleton pattern in Go, it's a topic of some debate.
+The Singleton pattern can introduce global state and make testing and concurrent programming more challenging.
+It can also make code tightly coupled and harder to maintain.
+In general, it's recommended to favor dependency injection and explicit collaboration over singletons.
+
+However, there might be scenarios where the Singleton pattern is still useful,
+such as managing shared resources or maintaining a global configuration.
+Just be mindful of the potential drawbacks and consider alternative patterns when appropriate.
+*/
+
+var (
+	forkManagerInstance     *forkManager
+	forkManagerInstanceLock sync.Mutex
+)
+
+type forkManager struct {
+	lock sync.Mutex
+
+	forkMap     map[ForkName]*Fork
+	handlersMap map[ForkHandlerName][]ForkActiveHandler
+}
+
+func GetInstance() *forkManager {
+	forkManagerInstanceLock.Lock()
+	defer forkManagerInstanceLock.Unlock()
+
+	if forkManagerInstance == nil {
+		forkManagerInstance = &forkManager{
+			forkMap:     map[ForkName]*Fork{},
+			handlersMap: map[ForkHandlerName][]ForkActiveHandler{},
+		}
+	}
+
+	return forkManagerInstance
+}
+
+func (fm *forkManager) RegisterFork(name ForkName) {
+	fm.lock.Lock()
+	defer fm.lock.Unlock()
+
+	fm.forkMap[name] = &Fork{
+		Name:            name,
+		FromBlockNumber: 0,
+		IsActive:        false,
+		Handlers:        map[ForkHandlerName]interface{}{},
+	}
+}
+
+func (fm *forkManager) RegisterHandler(forkName ForkName, handlerName ForkHandlerName, handler interface{}) error {
```

Comment:
Why this is not a part of `RegisterFork()`? I assume that when I register a fork I want to register all of its handlers at the same time.

Reply:
Because not every fork will have handlers for everything.
 Lets say there are forks A, B, C
 But it is possible to register "extra data" handlers only for forks A and C
 and to register "proposal selection" handlers for forks B and C